src/app/modules/spcp/sp.oidc.client.ts

```diff
@@ -0,0 +1,583 @@
+import axios from 'axios'
+import { ObjectId } from 'bson'
+import { createPrivateKey, createPublicKey, KeyObject } from 'crypto'
+import {
+  compactDecrypt,
+  JWTPayload,
+  jwtVerify,
+  JWTVerifyResult,
+  SignJWT,
+} from 'jose'
+import jwkToPem from 'jwk-to-pem'
+import NodeCache from 'node-cache'
+import { BaseClient, Issuer } from 'openid-client'
+import promiseRetry from 'promise-retry'
+
+import {
+  CreateAuthorisationUrlError,
+  CreateJwtError,
+  ExchangeAuthTokenError,
+  GetDecryptionKeyError,
+  GetVerificationKeyError,
+  InvalidIdTokenError,
+  JwkError,
+  MissingIdTokenError,
+  VerificationKeyError,
+} from './sp.oidc.client.errors'
+import {
+  CryptoKeySet,
+  isEC,
+  isECPrivate,
+  isSigningKey,
+  PublicJwks,
+  Refresh,
+  SecretJwks,
+  SigningKey,
+  SpOidcClientCacheConstructorParams,
+  SpOidcClientConstructorParams,
+} from './sp.oidc.client.types'
+
+/**
+ * Cache class which provides read-through capability and refresh-ahead before expiry
+ * Handles discovery and retrieval of NDI's public jwks
+ * Exported for testing
+ */
+export class SpOidcClientCache extends NodeCache {
+  #spOidcNdiDiscoveryEndpoint: string
+  #spOidcNdiJwksEndpoint: string
+  #spOidcRpClientId: string
+  #spOidcRpRedirectUrl: string
+  #spOidcRpSecretJwks: SecretJwks
+
+  /**
+   * Constructor for cache
+   * On instantiation, trigger a refresh to populate the cache
+   */
+  constructor({
+    options,
+    spOidcNdiDiscoveryEndpoint,
+    spOidcNdiJwksEndpoint,
+    spOidcRpClientId,
+    spOidcRpRedirectUrl,
+    spOidcRpSecretJwks,
+  }: SpOidcClientCacheConstructorParams) {
+    super(options)
+
+    this.#spOidcNdiDiscoveryEndpoint = spOidcNdiDiscoveryEndpoint
+    this.#spOidcNdiJwksEndpoint = spOidcNdiJwksEndpoint
+    this.#spOidcRpClientId = spOidcRpClientId
+    this.#spOidcRpRedirectUrl = spOidcRpRedirectUrl
+    this.#spOidcRpSecretJwks = spOidcRpSecretJwks
+
+    // On expiry, refresh cache. If fail to refresh, log but do not throw error.
+    this.on('expired', () =>
+      this.refresh().catch((err) =>
+        console.warn(
+          `Attempted but failed to refresh sp oidc cache on expiry. Reason: ${err}`,
+        ),
+      ),
+    )
+
+    // Trigger refresh on instantiation to populate cache. If fail to refresh, log but do not throw error.
+    void this.refresh().catch((err) =>
```

Comment:
> Ideally, the implementation should only have a single in-flight refresh per node process at a time

Can look into this

> The app is not considered ready till we have fetched the data

This approach might be problematic, because the data fetch could fail for multiple reasons (including downtime by NDI). Singpass integration is only one feature of the forms app, so we should not couple the availability of the entire form app to this. Hence, in the current implementation, if the data is not available (whether in cache or from NDI), an error is thrown and handled appropriately

Reply:
update: implemented single in-flight refresh per node process, by storing the promise on a refresh attempt, and returning that promise to future refresh attempts if the promise is still pending